src/typemap.c

```diff
@@ -445,68 +455,76 @@ int jl_typemap_visitor(union jl_typemap_t cache, jl_typemap_visitor_fptr fptr, v
+    return (jl_is_datatype(ty) && ((jl_datatype_t *)ty)->uid != 0 && !is_kind(ty));
 }
 
-static int jl_typemap_intersection_array_visitor(struct jl_ordereddict_t *a, jl_value_t *ty, int tparam,
-        int offs, struct typemap_intersection_env *closure)
+static int jl_typemap_intersection_array_visitor(struct jl_ordereddict_t *a, jl_value_t *ty,
+                                                 int tparam, int offs,
+                                                 struct typemap_intersection_env *closure)
 {
-    size_t i, l = jl_array_len(a->values);
-    union jl_typemap_t *data = (union jl_typemap_t*)jl_array_data(a->values);
+    size_t              i, l = jl_array_len(a->values);
+    union jl_typemap_t *data = (union jl_typemap_t *)jl_array_data(a->values);
     for (i = 0; i < l; i++) {
         union jl_typemap_t ml = data[i];
         if (ml.unknown == jl_nothing)
             continue;
         jl_value_t *t;
-        if (jl_typeof(ml.unknown) == (jl_value_t*)jl_typemap_level_type) {
+        if (jl_typeof(ml.unknown) == (jl_value_t *)jl_typemap_level_type) {
             t = ml.node->key;
         }
         else {
             t = jl_field_type(ml.leaf->sig, offs);
             if (tparam)
                 t = jl_tparam0(t);
         }
-        if (ty == (jl_value_t*)jl_any_type || // easy case: Any always matches
-            (tparam ?  // need to compute `ty <: Type{t}`
-             (jl_is_uniontype(ty) || // punt on Union{...} right now
-              jl_typeof(t) == ty || // deal with kinds (e.g. ty == DataType && t == Type{t})
-              (jl_is_type_type(ty) && (jl_is_typevar(jl_tparam0(ty)) ?
-                                       jl_subtype(t, ((jl_tvar_t*)jl_tparam0(ty))->ub, 0) : // deal with ty == Type{<:T}
-                                       jl_subtype(t, jl_tparam0(ty), 0)))) // deal with ty == Type{T{#<:T}}
-                    : jl_subtype(t, ty, 0))) // `t` is a leaftype, so intersection test becomes subtype
-            if (!jl_typemap_intersection_visitor(ml, offs+1, closure))
+        if (ty == (jl_value_t *)jl_any_type || // easy case: Any always matches
+            (tparam ?                          // need to compute `ty <: Type{t}`
+                 (jl_is_uniontype(ty) ||       // punt on Union{...} right now
+                  jl_typeof(t) ==
+                      ty || // deal with kinds (e.g. ty == DataType && t == Type{t})
+                  (jl_is_type_type(ty) &&
+                   (jl_is_typevar(jl_tparam0(ty))
+                        ? jl_subtype(t, ((jl_tvar_t *)jl_tparam0(ty))->ub, 0)
+                        :                                   // deal with ty == Type{<:T}
+                        jl_subtype(t, jl_tparam0(ty), 0)))) // deal with ty == Type{T{#<:T}}
+                    : jl_subtype(
+                          t, ty,
+                          0))) // `t` is a leaftype, so intersection test becomes subtype
```

Comment:
the original was pretty complex (and probably should be refactored), but this reformatting is much worse